flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java

```diff
@@ -152,37 +155,37 @@ public BarrierBuffer(InputGate inputGate, BufferBlocker bufferBlocker, long maxB
 		this.queuedBuffered = new ArrayDeque<BufferOrEventSequence>();
 	}
 
+	@Override
+	public CompletableFuture<?> isAvailable() {
+		if (currentBuffered == null) {
+			return inputGate.isAvailable();
```

Comment:
I think this is not ok, `InputGate#AVAILABLE` is currently not the same as `AsyncDataInput#AVAILABLE`, which messes up the checks for referencial equality. I guess you want to replace the `InputGate#AVAILABLE` with the instance from the interface.